Feat/json serialize or expression

jaimeferj opened the pull request:
<!--
Thanks for opening a pull request!
-->

<!-- In the case this PR will resolve an issue, please replace ${GITHUB_ISSUE_ID} below with the actual Github issue id. -->
Closes https://github.com/apache/iceberg-python/issues/2519

# Rationale for this change

## Are these changes tested?
yes

## Are there any user-facing changes?

<!-- In the case of user-facing changes, please add the changelog label. -->

jaimeferj commented on pyiceberg/expressions/__init__.py:
```diff
@@ -302,12 +302,19 @@ def __getnewargs__(self) -> Tuple[BooleanExpression, BooleanExpression]:
+    def __init__(self, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression) -> None:
+        if isinstance(self, Or) and not hasattr(self, "left") and not hasattr(self, "right"):
+            super().__init__(left=left, right=right)
+
```
This weird `__init__` is to make sure that if the class has been already initialized in `__new__`, it is not initialized again.
If this is not made, multiple things can happen:
- If *rest is given, although `_build_balanced_tree` build an `Or` instance correctly, Python would call again `Or` initializer with the original parameters `left` and `right` resulting in `Or(left, right)`, ignoring the created `Or` instance from `_build_balanced_tree`
- If left or right is `AlwaysTrue` it will get overriden and the instance will be initialized as `Or(left, right)`
- If left and right is `AlwaysFalse` it will get overriden and the instance will be initialized as `Or(left, right)`

I do not know why exactly but it gotta be something from Pydantic

jaimeferj replied:
I cannot test it now, but I am pretty sure that I left out `obj.left = left` because of how Python works with `__new__`. If you return a class of the same type that happens:

> If [__new__()](https://docs.python.org/3.12/reference/datamodel.html#object.__new__) is invoked during object construction and it returns an instance of cls, then the new instance’s [__init__()](https://docs.python.org/3.12/reference/datamodel.html#object.__init__) method will be invoked like __init__(self[, ...]), where self is the new instance and the remaining arguments are the same as were passed to the object constructor.
Ref: https://docs.python.org/3.12/reference/datamodel.html#object.__new__

So that is why I purposely left the object uninitialized, so that when the raw obj is returned it is initialized by `__init__`.

If the object has already being initialized by `_build_balanced_tree` I just leave it as is.

Fokko commented on pyiceberg/expressions/__init__.py:
```diff
@@ -302,12 +302,19 @@ def __getnewargs__(self) -> Tuple[BooleanExpression, BooleanExpression]:
 
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    type: TypingLiteral["str"] = Field(default="or", alias="type")
```
```suggestion
    type: TypingLiteral["or"] = Field(default="or", alias="type")
```

Fokko commented on tests/expressions/test_expressions.py:
```diff
@@ -702,6 +702,10 @@ def test_or() -> None:
 
+    assert (
+        or_.model_dump_json()
+        == '{"type":"or","left":"IsNull(term=Reference(name=\'a\'))","right":"IsNaN(term=Reference(name=\'b\'))"}'
```
This test is invalid, it should return something like:

```suggestion
        == '{"type":"or","left":{"type":"is-null","term":"a"},"right":{"type":"is-nan","term":"b"}}'
```

But the unary predicates are still WIP: https://github.com/apache/iceberg-python/pull/2598

Fokko replied:
I think a seperate test would be best :)

jaimeferj commented:
1. Rebased to main for the `LiteralPredicate` changes to make the test.
2. Added the new test.
3. Removed the serializers.
4. Tests and linting passing

jaimeferj commented:
CI failing at
```mypy.....................................................................Failed
- hook id: mypy
- exit code: 1
  tests/utils/test_manifest.py:51: error: Unused "type: ignore" comment  [unused-ignore]
  Found 1 error in 1 file (checked 166 source files)
```

PS: I see that you solved it in https://github.com/apache/iceberg-python/pull/2654

Fokko reviewed (approved):
LGTM! Thanks @jaimeferj

Fokko commented:
@jaimeferj Also feel free to fix the CI in this branch
